ocaml/libs/tracing/tracing_export.ml

```diff
@@ -0,0 +1,346 @@
+            ; duration: int
+            ; kind: string option
+            ; localEndpoint: zipkinEndpoint
+            ; annotations: annotation list
+            ; tags: (string * string) list
+          }
+          [@@deriving rpcty]
+
+          type t_list = t list [@@deriving rpcty]
+
+          let kind_to_zipkin_kind = function
+            | SpanKind.Internal ->
+                None
+            | k ->
+                Some k
+
+          let json_of_t_list s =
+            Rpcmarshal.marshal t_list.Rpc.Types.ty s |> Jsonrpc.to_string
+        end
+
+        let zipkin_span_of_span (s : Span.t) : ZipkinSpan.t =
+          let serviceName = get_service_name () in
+          let annotations =
+            s
+            |> Span.get_events
+            |> List.map (fun event : ZipkinSpan.annotation ->
+                   let timestamp =
+                     int_of_float (event.SpanEvent.time *. 1000000.)
+                   in
+                   let value = event.SpanEvent.name in
+                   {timestamp; value}
+               )
+          in
+          {
+            id= s |> Span.get_context |> SpanContext.span_id_of_span_context
+          ; traceId=
+              s |> Span.get_context |> SpanContext.trace_id_of_span_context
+          ; parentId=
+              s
+              |> Span.get_parent
+              |> Option.map (fun x ->
+                     x
+                     |> Span.get_context
+                     |> SpanContext.span_id_of_span_context
+                 )
+          ; name= s |> Span.get_name
+          ; timestamp= int_of_float (Span.get_begin_time s *. 1000000.)
+          ; duration=
+              Option.value (Span.get_end_time s)
+                ~default:(Unix.gettimeofday () *. 1000000.)
+              -. Span.get_begin_time s
+              |> ( *. ) 1000000.
+              |> int_of_float
+          ; kind=
+              s
+              |> Span.get_span_kind
+              |> ZipkinSpan.kind_to_zipkin_kind
+              |> Option.map SpanKind.to_string
+          ; localEndpoint= {serviceName}
+          ; annotations
+          ; tags=
+              Attributes.fold
+                (fun k v tags -> (k, v) :: tags)
+                (Span.get_attributes s) []
+          }
+
+        let content_of (spans : Span.t list) =
+          List.map zipkin_span_of_span spans |> ZipkinSpan.json_of_t_list
+      end
+    end
+  end
+
+  module Destination = struct
+    module File = struct
+      let trace_log_dir = ref "/var/log/dt/zipkinv2/json"
+
+      let max_file_size = ref (1 lsl 20)
+
+      let compress_tracing_files = ref true
+
+      let set_trace_log_dir dir = trace_log_dir := dir
+
+      let get_trace_log_dir () = !trace_log_dir
+
+      let set_max_file_size size = max_file_size := size
+
+      let set_compress_tracing_files enabled = compress_tracing_files := enabled
+
+      let file_name = ref None
+
+      let lock = Mutex.create ()
+
+      let new_file_name () =
```

Comment:
It would be more conventional to call this `make_file_name` or `create_file_name` to convey that this is not a pure function; I think updating `file_name` should be left to the caller. Alternatively, a `get_file_name` should manage creating and updating the file name together but the current split is not symmetric.